File: 3568-78-1372-longest-zigzag-path-in-a-binary-tree/3568-78-1372-longest-zigzag-path-in-a-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var longestZigZag = function(root) {
    let max = 0

    const helper = (root,dir,curr=0) => {
        if(!root)
            return
        max = Math.max(max,curr)
        if(dir === 'R'){
            helper(root.left,'L',1+curr)
            helper(root.right,'R',1)
        }else{
            helper(root.left,'L',1)
            helper(root.right,'R',1+curr)
        }
    }
    helper(root,'L')
    helper(root,'R')
    return max
};

File: 3568-78-1372-longest-zigzag-path-in-a-binary-tree/3568-78-1372-longest-zigzag-path-in-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int max = 0;
    public int longestZigZag(TreeNode root) {
        path(root,'L',0);
        path(root,'R',0);
        return max;
    }

    public void path(TreeNode root,char dir,int curr){
        if(root==null)
            return;
        max = Math.max(max,curr);
        if(dir=='L'){
            path(root.left,'L',1);
            path(root.right,'R',1+curr);
        }else{
            path(root.left,'L',1+curr);
            path(root.right,'R',1);
        }
    }
}